File: src/Runtime/Sentence/SentenceExpressionNot.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {
	public class SentenceExpressionNot : SentenceExpression {
		private SentenceExpression _expression = null;
		public SentenceExpressionNot(SentenceExpression expression) {
			_expression = expression;
		}
		public override ExpressionType GetExpressionType() { return ExpressionType.Not; }
		public override ExecuteResult Execute(Space space) {
			if (!IsSuccess(_expression.Execute(space))) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.ExpressionNot, "The [not-expression] execute failed!");
				return ExecuteResult.Failed;
			}

			var ret = ValueTool.ToLogic(_expression.value);
			SetValue(new ValueBool(!ret));

			return ExecuteResult.Successed;
		}
	}
} // namespace peak.interpreter

File: src/Runtime/Sentence/sentence_expression_not.cpp
#include "sentence_expression_not.h"
#include "runtime/value/value_tool.h"

using namespace peak;

SentenceExpressionNot::SentenceExpressionNot(std::shared_ptr<SentenceExpression> expression)
	: _expression(expression) {
}
ExecuteResult SentenceExpressionNot::Execute(std::shared_ptr<Space> space) {
	if (!IsSuccess(_expression->Execute(space))) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::ExpressionNot, "The [not-expression] execute failed!");
		return ExecuteResult::Failed;
	}

	auto ret = ValueTool::ToLogic(_expression->GetValue());
	SetValue(std::shared_ptr<Value>(new ValueBool(!ret)));

	return ExecuteResult::Successed;
}
